### ISSUE TITLE
Arabic Scripts

### ISSUE THREAD

**Issue by miracle2k** [ritzyed/ritzy#25]
I know RTL languages are currently not supported, but just as a note for the `isOpenTypeJsReliable` function: The OpenType char-by-char approach is not viable for some scripts (like arabic), because characters are different depending on whether they are standalone, or merged together.

Was the canvas measuring a speed concern, or why was the OpenType measuring mechanism chosen?

**Comment by rocketraman** [comment]
Yes, OpenType measuring was chosen for performance. Currently, both OpenType and the fallback canvas mechanism both use the char-by-char approach. However, I think it should be quite doable to use a multi-char measurement approach for canvas for sure.

For opentype, char-by-char could still work: it depends on what you're calling "merged characters" have a single unicode char value, and whether the font has a corresponding glyph for that char. Do you have an example?

**Comment by miracle2k** [comment]
The way it works is that a character looks different if it's at the beginning, the middle, the end, or standalone (sort of like longhand writing).

My understanding (I'm no expert) is the following: Each character is just a single unicode code point. The Font uses contextual OpenType features to choose what to render.

I see that CodeMirror uses the DOM to measure text. Is that something you have considered too?

https://github.com/codemirror/CodeMirror/blob/67365d2646551a0dcf9a18e2d1998686907233cd/lib/codemirror.js#L1849

**Comment by rocketraman** [comment]
> I see that CodeMirror uses the DOM to measure text. Is that something you have considered too?

I did consider that but a) I couldn't figure out the implementation :-), and b) I didn't spend too much time on it because I think it is a lot slower. For prosemirror that's probably ok but Ritzy needs these numbers a lot more often (because the surface layout is controlled completely by Ritzy). I could be wrong though. Probably worth trying out now that Prosemirror figured out the implementation.
